docs/RationaleFIRRTL.md

```diff
@@ -59,12 +59,55 @@ community that would benefit from adding support for these, we can do so.
+critical that Chisel, and by extension FIRRTL, have language-level semantics
+about how entities are named and how named entities are used and transformed.  
+This must specify which entities with names in Chisel generate predictable 
+output.  Since names serve multiple purposes in a design; for example, 
+debugging, test-bench attachment, hooks for physical layout, etc; we must 
+balance multiple needs.  This section describes the base semantics, which are 
+conservative and aimed at enabling debugging.  _It is expected that the 
+compiler provide flags to relax these rules to produce more synthesis- or 
+production-ready output_.
+
+Modules shall use the name given in Chisel, unless they conflict with a verilog 
+reserved word, not withstanding de-duplication or relevant annotations on the module.
+
+Instances shall use the name given in Chisel, unless they conflict with a 
+verilog reserved word.  Instances have preferential use of the name in the output in 
+case of a conflict, after ports.
+
+Chisel provides a "Don't Touch" annotation to protect entities from 
+transformation.  A don't touch on a wire or node produces a wire in verilog and 
+preserves the data-flow through that wire.  Even a wire driven by a constant 
+shall not have the constant forwarded around the wire.  This is because a 
+"don't touch" annotation signals the possible public use of a wire and one 
+common use is to provide a place to drive a new value into the logic from an 
+external test-bench.  If the node or wire is named (and it always should be for 
+Chisel don't touch), this name is used, unless it conflicts with a verilog 
+reserved word.  This wire has preferential use of the name in the output in 
+case of a conflict, after ports and instances.
+
+Named wires and nodes in FIRRTL shall appear as a wire in the output verilog.  
+There is no requirement that data-flow through a wire be maintained, only that 
+the data-flow into a wire be maintained.  This allows bypassing and forwarding 
+around wires who exist solely because of their name.  An implementation may 
+choose to not bypass trivial wires to reduce unused wire lint warnings, but 
+shouldn't cause other lint warnings to avoid unused wire warnings.  A named 
+wire without a symbol is thus equivalent to a named read-probe in the circuit.
+
+Any name of an entity inside a module which starts with `_` may be discarded.  
+This name pattern indicates the name is for convenience in the Chisel code 
+(often temporaries are required) and there is no expectation it exist in the 
+output.
+
+### Mandatory Renaming
+
+We want the naming of Verilog objects to match the names used in
+the original CHISEL, but in several passes, there is mandatory renaming.  It is 
+important that this be a predictable transformation.  For example, after 
+bundles are replaced with scalars in the lower-types pass, each field should be 
+prefixed with the bundle name:
```

Comment:
Eventually expanding on the _exact_ algorithm here would be good. Not blocking for this PR.